lib/parse/parser_interpreter.es6.js

```diff
@@ -0,0 +1,312 @@
+      // interpretation.
+      const ret = v => v === null ?
+            (this.delegate ?
+             this.delegate.interpretKeyValue('count', value, parser) :
+             this.interpretKeyword(`count:${value}`, parser)) : v;
+
+      const count = parseInt(value);
+      if (Number.isNaN(count)) return ret(null);
+
+      const cols = this.selected
+            .filter(col => this.CompatProperty.isInstance(col));
+      return this.EQ(this.ARRAY_COUNT(this.SEQ.apply(this, cols),
+                                      this.TRUTHY()),
+                     count);
+    },
+    function getMatchingColumns(key) {
+      // Match [browser name][browser version][os name][os version] where
+      // names are /[^0-9]+/ and versions are /[0-9.-]+/.
+      const match = key.match(/^([^0-9]+)([0-9.-]+)?([^0-9]+)?([0-9.-]+)?/i);
+      if (match === null) return [];
+
+      const browserName = match[1] && match[1].toLowerCase();
+      const browserVersion = match[2] && match[2].toLowerCase();
+      const osName = match[3] && match[3].toLowerCase();
+      const osVersion = match[4] && match[4].toLowerCase();
+
+      const columns = this.selectable
+            .filter(col => this.CompatProperty.isInstance(col));
+
+      return columns.filter(
+          c => (!browserName || c.release.browserName.toLowerCase()
+                .startsWith(browserName)) &&
+            (!browserVersion || c.release.browserVersion.toLowerCase()
+             .startsWith(browserVersion)) &&
+            (!osName || c.release.osName.toLowerCase()
+             .startsWith(osName)) &&
+            (!osVersion || c.release.osVersion.toLowerCase()
+             .startsWith(osVersion)));
+    },
+  ],
+});
+
+foam.CLASS({
+  name: 'QueryParser',
+  package: 'org.chromium.parse',
+  implements: ['foam.mlang.Expressions'],
+
+  requires: [
+    'foam.core.Property',
+    'foam.core.String',
+    'org.chromium.parse.DefaultQueryInterpreter',
+    'org.chromium.parse.ReleaseApiQueryInterpreter',
+  ],
+
+  properties: [
+    {
+      class: 'Class',
+      name: 'of',
+      documentation: 'The class of items in the DAO being queried.',
+    },
+    {
+      class: 'FObjectProperty',
+      of: 'org.chromium.parse.QueryInterpreter',
+      name: 'interpreter',
+      factory: function() {
+        return this.ReleaseApiQueryInterpreter.create({
+          delegate: this.DefaultQueryInterpreter.create(),
+        });
+      },
+    },
+  ],
+
+  methods: [
+    function parseString(str, opt_name) {
+      // Parse string; on complete parse, return result.
+      const result = this.queryGrammar.getParse(str, opt_name);
+      if (result && result.pos === str.length) return result.value;
+
+      // Store partial parse, and extract keywords from remaining string.
+      const partialParse = result && result.value;
+      const keywords = str.substr(partialParse ? result.pos : 0)
+            .replace(/[ \t\r\n]+/g, ' ').split(' ').filter(kw => !!kw);
```

Comment:
From offline: filter could be replaced with trim to make it more explicit that the goal is to remove empty strings, since those empty strings can only be from leading and trailing whitespace after the collapsing done by the replace call.